File: linda-presentation/src/main/java/it/unibo/coordination/linda/presentation/RegexTemplateDeserializer.java
package it.unibo.coordination.linda.presentation;

import com.fasterxml.jackson.databind.ObjectMapper;
import it.unibo.coordination.linda.text.RegexTemplate;

import java.util.Map;

class RegexTemplateDeserializer extends DynamicDeserializer<RegexTemplate> {

    public RegexTemplateDeserializer(MIMETypes mimeType, ObjectMapper mapper) {
        super(RegexTemplate.class, mimeType, mapper);
    }

    @Override
    public RegexTemplate fromDynamicObject(Object dynamicObject) {
        if (dynamicObject instanceof Map) {
            Map<String, ?> map = (Map<String, ?>) dynamicObject;
            if (map.containsKey("template")) {
                Object tuple = map.get("template");
                if (tuple instanceof String) {
                    return RegexTemplate.of((String) tuple);
                }
            }
        }
        throw new IllegalArgumentException("Cannot read " + getSupportedMIMEType());
    }
}

File: linda-text-presentation/src/main/kotlin/it/unibo/coordination/linda/text/RegexTemplateDeserializer.kt
package it.unibo.coordination.linda.text

import com.fasterxml.jackson.databind.ObjectMapper
import it.unibo.presentation.DynamicDeserializer
import it.unibo.presentation.MIMETypes

internal class RegexTemplateDeserializer(mimeType: MIMETypes, mapper: ObjectMapper) : DynamicDeserializer<RegexTemplate>(RegexTemplate::class.java, mimeType, mapper) {
    override fun fromDynamicObject(dynamicObject: Any): RegexTemplate {
        if (dynamicObject is Map<*, *>) {
            if (dynamicObject.containsKey("template")) {
                val tuple = dynamicObject["template"]
                if (tuple is String) {
                    return RegexTemplate.of(tuple)
                }
            }
        }
        throw IllegalArgumentException("Cannot read $supportedMIMEType")
    }
}